fix(renderer): fix missing slot ref callback handling

yigityuce opened the pull request:
fixes one of the issues raised in #5335

adds a new internal variable which is stored during vnode rendering to reuse it within the relocatedNodes logic

<!-- Please refer to our contributing documentation for any questions on submitting a pull request, or let us know here if you need any help: https://github.com/ionic-team/stencil/blob/main/CONTRIBUTING.md -->


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Slot element's `ref` attribute is not handled correctly.

GitHub Issue Number:  #5335 


## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->
Fixed


## Documentation

<!-- Please add any link(s) to documentation-related pull requests here -->

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

## Testing

<!-- Please describe the steps you took to test the changes in this PR. These steps can be programmatic (e.g. unit tests) and/or manual. -->

## Other information

<!-- Any other information that is important to this PR such as screenshots of how a component looks before and after the change. -->

christian-bromann reviewed (changes requested):
Thanks for the contribution @yigityuce 🙏 

Can I ask you do to add an end-to-end test for this to ensure we don't regress on this? I recommend to look into `test/karma/test-app` and add a new `slot-XXX` folder and copy your example from the reproduction case.

Let me know if you have any questions or if I can assist you.

yigityuce commented:
Hi @christian-bromann thanks for the detailed explanation. I just updated the tests. 

However, even in the main branch 3 of 156 test cases are failing. It would be worth checking. I added my test and it seems it works. I know a bit about the TDD. So initially, I implemented the failing test case and then corrected it. So I am pretty sure it is passing 😃 

![image](https://github.com/ionic-team/stencil/assets/5856230/801fac7d-0e7e-412c-83a1-f71b3a2b7837)


One more question for future contributions. I don't have that much experience with karma, unfortunately. And as of now, I have done the following steps to have a working testing environment. Please confirm that the steps are ok and please provide a better and easier way to accomplish this such as starting it in watch mode aor sth like that.


```sh
cd test/karma
npm i

# npm run karma # failed

npm run karma.prod

```

christian-bromann commented:
> However, even in the main branch 3 of 156 test cases are failing. It would be worth checking.

Make sure you have the project compile (`npm run build`) and run `npm run test.karma.prod`. This also ensures that all example projects are compiled with the Stencil build you have.

> Please confirm that the steps are ok and please describe a way to make this operation more easier such as starting it in watch mode or sth like that.

As mentioned above the best way to start the e2e tests are by running `npm run test.karma.prod`. We are currently in the process to migrate over to WebdriverIO for browser based unit tests. In fact we start with this work next sprint (next week). We will make sure to document the process accordingly but I am sure it will become simpler moving on 😉

christian-bromann commented on test/karma/test-app/slot-ref/cmp.tsx:
```diff
@@ -0,0 +1,18 @@
+  render() {
+    return (
+      <Host>
+        <slot name="title" ref={() => (this.hostElement.setAttribute('data-ref', 'called'))} />
```
Can we update the test and use the provided element reference to verify that it was provided in the callback?

yigityuce replied:
done 🚀

christian-bromann reviewed (commented):
One little ask to extend the coverage of given tests.

christian-bromann reviewed (approved):
Perfect 👌 this LGTM 👍 

Let's have someone else from the @ionic-team/stencil review this.

yigityuce commented:
hey @christian-bromann do I need to do sth else to move forward?

rwaskiewicz commented:
@yigityuce Can you run `npm run format` when you get a chance on this branch? It looks like that CI check is currently failing

christian-bromann commented:
This has been released with Stencil 🚞 [v4.13.0](https://github.com/ionic-team/stencil/releases/tag/v4.13.0)
